gwu-libraries/scholarspace-hyrax
# T458 license field display

### Workflow notes

This branch has been rebased with #300 (PR #555), so the latter should be merged first.  

### Changes

- Updated `GwWork` and `GwEtd` views to include the license field presenter (making the field visible).
- Updated the authorities file (`license.yml`):
   - The deprecated Europeana license value has been set to inactive.
   - A new `All rights reserved` value has been added, which will display as plain text.
- Fixed a Hyrax [bug](https://github.com/samvera/hyrax/issues/5316) whereby works bearing an inactive authority value are not editable. 
- Added a rake task to update all works in the repository bearing the deprecated license value, replacing it with the new value.
- Added a spec test to validate the visibility of the license field.
- Minor spec changes:
   - Modified the Bulkrax importer spec to use the `TEMP_FILE_BASE` environment variable, if present, instead of defaulting to `scholarspace-hyrax/tmp`. 
   - Added values for all required fields to the `gw_work` factory, and added logic associating a user's abilities with that work. (See note on test coverage below.)

### Reviewer instructions

1. Follow setup instructions on #555 (if you haven't already done so).
2. Verify that all spec tests run correctly.
3. Verify that on visiting a work's page in the UI, the license field is visible.
4. Edit a work in the UI and verify that you can change the value in the license field and save the work.
5. Work through the steps in #555 to perform a Bulkrax import, but prior to uploading, edit the `metadata.csv` file, adding a `license` column and populating this column with the value `All rights reserved`. 
6. Confirm that the works created by the Bulkrax importer have the correct value in the license field.
7. Create some works with the deprecated license value (which will appear in the form as `All rights reserved (deprecated)`, and then run `bundle exec rake gwss:replace_license_value` (in the appropriate environment). Verify that the license values have been updated for those works to the non-deprecated `All rights reserved`.

### Test coverage

The current spec tests on this branch verify the visibility of the license field. I attempted to write a test for the `Edit` functionality in the UI, but I can't seem to get that working. A logged in `content-admin` user can access the `Edit` form for the work and click the `Save changes` button, but a) the changes don't seem to persist, and b) the work then becomes inaccessible to the user, even though the visibility on the work itself is still `open`. 

I can't tell if this stems from a Capybara issue, or from an issue with how we're creating test works. I adapted code from the `create_dummy_works` Rake task for adding a user's ability to a work; this step seemed necessary in order for the logged-in user to be able to access the `Edit` page. But I note that the works created with FactoryBot lack a value for the `state` field, which normally would contain a reference to a Fedora object -- could this be a problem? 

I haven't tried creating the work through Capybara actions; if the problem lies with FactoryBot, presumably, this approach would fix it. I'm happy to keep working on this, but perhaps @alepbloyd and I can coordinate efforts.
